design/src/docs/pokemon_service.md

```diff
@@ -0,0 +1,239 @@
+        .layer(AddExtensionLayer::new(shared_state)),
+);
+```
+The service is run by `hyper`:
+```rust
+hyper::Server::bind(&bind).serve(app.into_make_service());
+```
+
+
+### Folder structure
+Before describing how shapes are generated, it is helpful to know the folder structure of smithy-rs.
+Code generation happens in:
+* `/codegen`: it contains shared code for both client and server, but only generates a client
+* `/codegen-server`: server only. This project started with `codegen` to generate a client, but client and server share common code; that code lives in `codegen`, which `codegen-server` depends on
+* `/aws`: the AWS SDK, it deals with AWS services specifically. The folder structure reflects the project's, with the `rust-runtime` and the `codegen`
+* `/rust-runtime`: the generated client and server crates may depend on crates in this folder. Crates here are not code generated
+
+`/rust-runtime` crates ("runtime crates") are added to a crate's dependency only when used. If a model uses event streams, it will depend on `aws-smithy-eventstream`.
+
+### Generating code
+`smithy-rs` is a gradle plugin, not a command. Its entry point is in [RustCodegenPlugin::execute](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/RustCodegenPlugin.kt#L34) and
+inherits from `SmithyBuildPlugin` in [smithy-build](https://github.com/awslabs/smithy/tree/main/smithy-build). Code generation is in Kotlin and shared common, non-Rust specific code with the `smithy` Java repository.
+
+The comment at the beginning of `execute` described what a `Decorator` is and uses the following terms:
+* Context: contains the model being generated, projection and settings for the build
+* Decorator: (also referred to as customizations) customizes how code is being generated. AWS services are required to sign with the SigV4 protocol, and [a decorator](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4SigningDecorator.kt#L45) adds Rust code to sign requests and responses.
+Decorators are applied in reverse order of being added and have a priority order.
+* Writer: creates files and adds content; it supports templating, using `#` for substitutions
+* Location: the file where a symbol will be written to
+
+The only task of a `RustCodegenPlugin` is to construct a `CodegenVisitor` and call its [execute()](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/CodegenVisitor.kt#L115-L115) method.
+
+`CodegenVisitor::execute()` is given a `Context`, constructs the sequence of decorators and calls a [CodegenVisitor](https://github.com/awslabs/smithy-rs/blob/db48039065bec890ef387385773b37154b555b14/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/CodegenVisitor.kt#L44).
```

Comment:
No, the visitor does not construct the decorators. It already takes them as input:

https://github.com/awslabs/smithy-rs/blob/9ee45bfd4410e2c959a04dfe71ff727cf93975f2/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ServerCodegenVisitor.kt#L50-L53